Implemented API rate limiting

maximmasiutin opened the pull request:
- Added code to avoid triggering API rate limit (503 errors).
- Added new configuration parameter api_request_threshold with default 0.5 seconds; the value is configurable in the ini file without currently an intent of modifying it via the GUI as the limit is currently fixed at the server side, see https://rutracker.org/forum/viewtopic.php?p=75760916#75760916
- Added notice/debug log entries before API calls.
- Removed the implicit parameters to allow future parameter expansion (updated API calls to specify search mode explicitly.
- Removed default value for $timeout parameter in createApiClient method, making it mandatory).

maximmasiutin commented:
It was not logged how many side files are stored. How this value is logged as follows:
```
[2025-05-11T21:29:53.439582+03:00] webtlo.NOTICE: Хранится много сторонних раздач (2118 шт.). Рекомендуется отключить поиск сторонних раздач или добавить подразделы в хранимые. [] {"memory_usage":"50 MB","memory_peak_usage":"58 MB"}
```
and
```
[2025-05-11T21:29:53.439776+03:00] webtlo.DEBUG: Получаем сведения о сторонних раздачах (2118 шт.) по их хешам через API... [] {"memory_usage":"50 MB","memory_peak_usage":"58 MB"}
```

Besides that, there was a rate limit (503) error triggered on such requests; now the rate limit is honored and the rate limit error is does not happen.

maximmasiutin commented:
@NeBioNik - please review

NeBioNik commented on src/back/Update/TorrentsClients.php:
```diff
@@ -216,11 +216,18 @@ private function updateUntracked(bool $doUpdateUntracked): void
                     $this->logger->notice(
-                        'Хранится много сторонних раздач. Рекомендуется отключить поиск сторонних раздач или добавить подразделы в хранимые.'
+                        'Хранится много сторонних раздач ({count} шт.). Рекомендуется отключить поиск сторонних раздач или добавить подразделы в хранимые.',
+                        ['count' => $countUntracked]
```
What is the point of this amendment if the number of topics is displayed literally one line higher?
```
12.05.2025 15:08:06 INFO: Найдено уникальных сторонних раздач в клиентах: 8750 шт.
12.05.2025 15:08:06 NOTICE: Хранится много сторонних раздач. Рекомендуется отключить поиск сторонних раздач или добавить подразделы в хранимые.
```

maximmasiutin replied:
It was not obvious of what is considered "many", the line above mentioned only about the "unique" ones found lately, and it was not clear whether the "many" number considered those not unique; it might seem that the total stored number is higher than the newly found unique ones. These two log lines were perplexing before I added the number.

NeBioNik commented on src/back/Update/TorrentsClients.php:
```diff
@@ -216,11 +216,18 @@ private function updateUntracked(bool $doUpdateUntracked): void
                 }
 
                 // Пробуем найти в API раздачи по их хешам из клиента.
+
```
Same question.

maximmasiutin replied:
I agree that this line is not needed.

NeBioNik commented on src/back/Update/TopicsDetails.php:
```diff
@@ -65,6 +66,10 @@ private function fillDetails(int $beforeUpdate, int $perRun): void
+                $this->logger->debug(
+                    'Запрашиваем сведения о раздачах ({count} шт.) через API (запрос {run}/{runs})...',
+                    ['count' => count($topics), 'run' => $run, 'runs' => $runs]
+                );
```
What is the point of this amendment if the number of topics is displayed literally one line lower?

```
07.03.2025 08:34:43 DEBUG: Обновление раздач [19500/57279]
07.03.2025 08:34:51 DEBUG: Обновление раздач [21000/57279]
07.03.2025 08:34:59 DEBUG: Обновление раздач [22500/57279]
07.03.2025 08:35:06 DEBUG: Обновление раздач [24000/57279]
```

maximmasiutin replied:
I agree, this is not needed, thank you!

NeBioNik commented on src/back/Update/HighPriority.php:
```diff
@@ -123,6 +124,8 @@ public function update(): void
     private function processSubsectionTopics(array $topics, callable $avgProcessor): void
     {
+        $config = $this->settings->get();
+
```
For what?

maximmasiutin replied:
Sorry, the line is not needed. I thought I would take API minimum request threshold from the config, but there is another way how this setting is used.

NeBioNik commented on src/back/Update/HighPriority.php:
```diff
@@ -195,8 +198,12 @@ private function processSubsectionTopics(array $topics, callable $avgProcessor):
-                // Получить описание новых раздач.
-                $response = $this->apiClient->getTopicsDetails(array_keys($topicsInsert));
+                $this->logger->notice(
+                    'Получаем дополнительные сведения для новых раздач ({count} шт.) через API...',
```
This note will be performed in a cycle for 18284 topics, for groups of 500. Are you sure that 40 notice is a good idea? What is the purpose of this information and why is it in the logs?

maximmasiutin replied:
I agree with your objection; moreover, this line never appeared in my logs, doesn't seem typical for my workflows.

NeBioNik commented on src/back/External/Api/Actions/TopicsDetails.php:
```diff
@@ -26,24 +26,35 @@ trait TopicsDetails
+                        }
+                        $lastRequestTime = microtime(true);
+                    }
+
```
503 code processing is done in getRetryMiddleware. 
If the API responds with an error, the request will be repeated later automatically. 

Have you compared the update time of a large number of topics in its current form and with the fixes made? What is the point of them?

https://github.com/keepers-team/webtlo/blob/master/src/back/External/Shared/RetryMiddleware.php#L27-L62
https://github.com/caseyamcl/guzzle_retry_middleware?tab=readme-ov-file#usage

NeBioNik replied:
Also, `Pool` used `concurrency=4`. In my testing I got 503 0 times.

If you want to see some logs, just add here
https://github.com/keepers-team/webtlo/blob/master/src/back/External/Api/Actions/TopicsDetails.php#L110
line like this
```php
$logger->debug('Done chunk request', ['index' => $index]);
```

maximmasiutin replied:
The point of this change is to avoid the 503 code at all - we should not make requests to the API with the frequency higher than the rate limit. I was getting numerous 503 errors before I implemented this change.
[503.zip](https://github.com/user-attachments/files/20166037/503.zip)

NeBioNik replied:
Just change concurrency to 2.

maximmasiutin commented:
These changes are substantial:
- Added code to avoid triggering API rate limit (503 errors).
- Added new configuration parameter api_request_threshold with default 0.5 seconds; the value is configurable in the ini file without currently an intent of modifying it via the GUI as the limit is currently fixed at the server side, see https://rutracker.org/forum/viewtopic.php?p=75760916#75760916 (the value of 0 removes the throttling to mimic old behavior)

These changes are optional (cosmetic) and are not actually needed:
- Added notice/debug log entries before API calls.
- Removed the implicit parameters to allow future parameter expansion (updated API calls to specify search mode explicitly.
- Removed default value for $timeout parameter in createApiClient method, making it mandatory).

Please let me know whether you are OK with the substantial changes to avoid triggering the rate limit by sending the API with frequency higher than needed.

NeBioNik commented:
So. I have done a lot of testing on this problem, which I personally only manage to replicate if I specify `concurrency=25`.

In an ideal world, 50 requests (32 * 50 topics) are executed in 50/2 = 25 seconds. 

At `concurrency=2` - this is true 
At `concurrency=4` - this is true
At `concurrency=16` - this is true 
At `concurrency=25` - this is true

With `concurrency=25` - this is ALSO true, but I got about 10 503 errors.

If you add some algorithm to the formula that would prohibit sending more than 2 requests per second with sleep (which stops php completely and the meaning of asynchronous requests disappears), the execution time increases by a couple of seconds: 

At `concurrency=2` - 25s, no delays 
At `concurrency=4` - 25.5s with some delays of ~0.5s 
At `concurrency=16` - 32s with delays of 0.5-1s, no 503 errors 
At `concurrency=25` - 37s with delays of 0.5-1s, with 3-5 503 errors 

From all of the above, I conclude that the introduction of a script execution pause in asynchronous queries does more harm (increasing execution time) than good (reducing the number of 503 errors).

There is no sense to change anything, the whole problem is solved by changing concurrency from 4 to 2, which I have done here https://github.com/keepers-team/webtlo/commit/26f6e8ab1a728855fc5cd8d830f720eae0882412 

Maybe I'll add some algorithm that doesn't stop the code using [React\Async\await](https://github.com/reactphp/async), but later.

This PR is closed.

> These changes are optional (cosmetic) and are not actually needed:
> 
> * Added notice/debug log entries before API calls.
> * Removed the implicit parameters to allow future parameter expansion (updated API calls to specify search mode explicitly.
> * Removed default value for $timeout parameter in createApiClient method, making it mandatory).

I recommend that these changes be put in a separate PR fixing the problems previously discussed.

maximmasiutin commented:
One API request takes ~80 ms from my computer.

`curl -s -w "%{time_total}\n" -o /dev/null "https://api.rutracker.cc/v1/get_tor_topic_data?by=topic_id&val=1%2C2142"`
0.084856

So, with concurrency=1 you could theoretically get ~12 requests per second unless you are throttled.

But the API seems to impose throttling by itself. I run curl in a loop and only the first request is fast, subsequent requests are indeed throttled by 2 requests per second; however, the codes are 200, not 503.
```
0.083114
0.457092
0.493778
0.492836
0.492887
0.494634
0.493070
0.496592
0.491608
0.494075
```

Anyway, if the server is imposing the limit, it it better to not provoke the 503 errors: if network latency allows us to make 12 requests per second when the rule permits 2, we should obey the rule.
